### PR TITLE
Internal: Fix Data/V2  force `$callback` to be `base_callback`. [ED-5293]

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
PHP-Lint

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]

File: docs/data/base/v2/controller.md [hunk 1]
```diff
@@ -13,8 +13,9 @@
...
 	| `Controller::get_controller_index()`    |                                | `WP_REST_Response` or `WP_Error`               | Retrieves rest route(s) index for current controller.
-	| `Controller::get_items_args()`          | `string` $methods              | `WP_REST_Response` or `WP_Error`               | Retrieves rest route(s) index for current controller.
-	| `Controller::get_permission_callback()` |  `WP_REST_Request` $request    | `{array}`                                      | Return args of of items route for each method, (Default args for `Base\Endpoint\Index`). 
+	| `Controller::get_items_args()`          | `string` $methods              | `Array`                                        | Get items args of index endpoint to be registered, for `GET` by default will return `$this->get_collection_params()` 
+	| `Controller::get_item_args()`           | `string` $methods              | `Array`                                        | Get item args of index endpoint to be registered.
+	| `Controller::get_permission_callback()` |  `WP_REST_Request` $request    | `Boolean`                                      | Retrieves the state of should this request allow to run.
```
Comment (iNewLegend) [inline docs/data/base/v2/controller.md:18]:
```suggestion
	| `Controller::get_permission_callback()` |  `WP_REST_Request` $request    | `Boolean`                                      | Retrieves the state of should this request allow to run.
```
```suggestion
	| `Controller::get_permission_callback()` |  `WP_REST_Request` $request    | `Boolean`                                      | Retrieves the permission status for each request, `true` allow running.
```

File: data/v2/base/controller.php [hunk 2]
```diff
@@ -1,6 +1,8 @@
 <?php
+
```
Comment (iNewLegend) [inline data/v2/base/controller.php:2]:
```suggestion
```

File: data/v2/base/base-route.php [hunk 3]
```diff
@@ -218,13 +218,29 @@ protected function register_route( $route = '', $methods = WP_REST_Server::READA
...
+
 		return register_rest_route( $this->controller->get_namespace(), $route, [
 			[
 				'args' => $args,
 				'methods' => $methods,
 				'callback' => $callback,
-				'permission_callback' => function ( $request ) {
-					return $this->get_permission_callback( $request );
+				'permission_callback' => function () {
+					return true;
```
[the remark this answers is not in the document]
Comment (matipojo) [reply data/v2/base/base-route.php:244]: I didn't understand. if the real check is not here - remove it.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: data/v2/base/base-route.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (iNewLegend) [reply data/v2/base/base-route.php]: Handled.

File: data/v2/base/base-route.php [hunk 4]
```diff
@@ -218,13 +218,29 @@ protected function register_route( $route = '', $methods = WP_REST_Server::READA
 			'route' => $route,
 		];
 
+		/**
+		 * Determine behaviour of `base_callback()` and `get_permission_callback()`
+		 * For calling the right method, case for each situation:
+		 * like `get_item()` or `get_items()` and such for `base_callback()`.
+		 * and `get_item_permission_callback()` or `get_items_permission_callback()` and such for `get_permission_callback()`.
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply data/v2/base/base-route.php]: Handled.

File: data/v2/base/base-route.php [hunk 5]
```diff
@@ -221,9 +221,7 @@ protected function register_route( $route = '', $methods = WP_REST_Server::READA
 		/**
 		 * Determine behaviour of `base_callback()` and `get_permission_callback()`:
 		 * For `base_callback()` which applying the action.
-		 * if it's a one item request will call 'get_item(),create_item(),update_item(), delete_item().).
-		 * if it's a multi item request will call 'get_items(),create_items(),update_items(), delete_items().).
-		 * The `get_permission_callback()` also effected in the same terms eg: `get_item_permission_callback()` or `get_items_permission_callback()`.
+		 * Iif it's a one item request and should call `get_item_permission_callback()` or it's mutil items request and should call `get_items_permission_callback()`.
```
Comment (matipojo) [inline data/v2/base/base-route.php:224]:
```suggestion
		 * Whether it's a one item request and should call `get_item_permission_callback()` or it's mutil items request and should call `get_items_permission_callback()`.
```

File: tests/phpunit/elementor/data/v2/base/test-base-route.php [hunk 6]
```diff
@@ -86,11 +87,32 @@ public function test_get_base_route__from_grandchild_endpoint() {
...
+
+		// By pass original permission callback.
+		$endpoint->bypass_original_permission( true );
+		$endpoint->bypass_set_value( false );
+
+		// Act.
+		$data = $this->manager->run_endpoint( $endpoint->get_base_route() );
+
+		// Assert.
+		$this->assertEquals( 'rest_forbidden', $data['code'] );
```
Comment (matipojo) [inline tests/phpunit/elementor/data/v2/base/test-base-route.php:114]: Why it's not an HTTP code?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/data/v2/base/test-base-route.php (hunk 6 above)
Comment (iNewLegend) [reply tests/phpunit/elementor/data/v2/base/test-base-route.php:114]: Handled.

**Review by matipojo — APPROVED** [review]

**Pull request merged**